src/debug-generator/debug-generator.c

```diff
@@ -20,20 +22,95 @@
 #include "unit-file.h"
 #include "unit-name.h"
 
+typedef enum BreakpointType {
+        BREAKPOINT_PRE_UDEV,
+        BREAKPOINT_PRE_SYSROOT_MOUNT,
+        BREAKPOINT_PRE_SWITCH_ROOT,
+        _BREAKPOINT_TYPE_MAX,
+        _BREAKPOINT_TYPE_INVALID = -EINVAL,
+} BreakpointType;
+
+typedef struct BreakpointInfo {
+        BreakpointType type;
+        const char *name;
+        const char *unit;
+        bool valid_in_initrd;
+        bool valid_in_main_system;
+} BreakpointInfo;
+
 static const char *arg_dest = NULL;
 static char *arg_default_unit = NULL;
 static char **arg_mask = NULL;
 static char **arg_wants = NULL;
 static bool arg_debug_shell = false;
 static char *arg_debug_tty = NULL;
 static char *arg_default_debug_tty = NULL;
+static uint32_t arg_breakpoints = 0;
 
 STATIC_DESTRUCTOR_REGISTER(arg_default_unit, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_mask, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_wants, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_debug_tty, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_default_debug_tty, freep);
 
+static const struct BreakpointInfo breakpoint_info_table[_BREAKPOINT_TYPE_MAX] = {
+        { BREAKPOINT_PRE_UDEV,          "pre-udev",        "breakpoint-pre-udev.service",        true,  true  },
+        { BREAKPOINT_PRE_SYSROOT_MOUNT, "pre-mount",       "breakpoint-pre-mount.service",       true,  false },
+        { BREAKPOINT_PRE_SWITCH_ROOT,   "pre-switch-root", "breakpoint-pre-switch-root.service", true,  false },
```

Comment:
the two boolean columns have double whitespace between them.

i am not a fan of showing the last two fields in such an opaque fashion here, without any immediate hint what they encode. the other three fields are somewhat self explanatory.

maybe add a comment as first row or so, which lists the field names, i.e.

```
/* ↓ type                  ↓ name     ↓ unit     ↓ valid_in_initrd  ↓ valid_in_main_system */
```

Reply:
an alternative would be two introduce:

```c
typedef enum BreakpointValidity {
        BREAKPOINT_IN_INITRD = 1 << 0,
        BREAKPOINT_ON_HOST   = 1 << 1,
} BreakpointValidity;
```

so that you can specify a single mask field instead of two booleans. This would become a lot more readable that way, hence much preferred I think.